WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCaseTest.kt

```diff
@@ -0,0 +1,112 @@
+
+import android.arch.core.executor.testing.InstantTaskExecutorRule
+import com.nhaarman.mockito_kotlin.any
+import com.nhaarman.mockito_kotlin.verify
+import com.nhaarman.mockito_kotlin.whenever
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.ArgumentCaptor
+import org.mockito.Mock
+import org.mockito.junit.MockitoJUnitRunner
+import org.wordpress.android.fluxc.Dispatcher
+import org.wordpress.android.fluxc.action.SiteAction
+import org.wordpress.android.fluxc.annotations.action.Action
+import org.wordpress.android.fluxc.store.SiteStore
+import org.wordpress.android.fluxc.store.SiteStore.NewSitePayload
+import org.wordpress.android.fluxc.store.SiteStore.OnNewSiteCreated
+import org.wordpress.android.fluxc.store.SiteStore.SiteVisibility
+import org.wordpress.android.test
+
+private val DUMMY_SITE_DATA: NewSiteCreationServiceData = NewSiteCreationServiceData(
+        123,
+        999,
+        "title",
+        "tagLine",
+        "slug"
+)
+private const val LANGUAGE_ID = "lang_id"
+
+@RunWith(MockitoJUnitRunner::class)
+class CreateSiteUseCaseTest {
+    @Rule
+    @JvmField val rule = InstantTaskExecutorRule()
+
+    @Mock lateinit var dispatcher: Dispatcher
+    @Mock lateinit var store: SiteStore
+    private lateinit var useCase: CreateSiteUseCase
+    private lateinit var event: OnNewSiteCreated
+
+    @Before
+    fun setUp() {
+        useCase = CreateSiteUseCase(dispatcher, store)
+        event = OnNewSiteCreated()
+        event.newSiteRemoteId = 123
+    }
+
+    @Test
+    fun coroutineResumedWhenResultEventDispatched() = test {
+        whenever(dispatcher.dispatch(any())).then { useCase.onNewSiteCreated(event) }
+        val resultEvent = useCase.createSite(DUMMY_SITE_DATA, LANGUAGE_ID)
+
+        assertThat(resultEvent).isEqualTo(event)
+    }
+
+    @Test
+    fun verifySiteDataPropagated() = test {
+        whenever(dispatcher.dispatch(any())).then { useCase.onNewSiteCreated(event) }
+        useCase.createSite(DUMMY_SITE_DATA, LANGUAGE_ID)
+
+        val captor = ArgumentCaptor.forClass(Action::class.java)
+        verify(dispatcher).dispatch(captor.capture())
+
+        assertThat(captor.value.type).isEqualTo(SiteAction.CREATE_NEW_SITE)
+        assertThat(captor.value.payload).isInstanceOf(NewSitePayload::class.java)
+        val payload = captor.value.payload as NewSitePayload
+        assertThat(payload.siteName).isEqualTo(DUMMY_SITE_DATA.siteSlug)
+        assertThat(payload.siteTitle).isEqualTo(DUMMY_SITE_DATA.siteTitle)
+        // TODO uncomment when the API is ready
+//        assertThat(payload.segmentId).isEqualTo(DUMMY_SITE_DATA.segmentId)
+//        assertThat(payload.verticalId).isEqualTo(DUMMY_SITE_DATA.verticalId)
+//        assertThat(payload.tagLine).isEqualTo(DUMMY_SITE_DATA.siteTagLine)
+    }
+
+    @Test
+    fun verifyDryRunIsFalse() = test {
```

Comment:
> It's funny that you brought this up on this specific method. I think I don't usually write these types of methods - tbh I think this is my first in WPAndroid.

This might not have been the best test to bring it up, I probably should have done it in the verticals PR, because even now when I look at the tests for the Verticals VM, I get really overwhelmed and don't know what's going on. I guess I'll have a better understanding of it when I implement the domains VM unit tests and will be interesting choice between following the verticals example and taking a different approach.

> We are giving control over those attributes to the clients even though I don't think they should ever change them. Wdyt?

I don't fully agree with this. These are fully reusable components, a different part of the app could have an entirely different set of requirements. I'd say if it's in the payload, the use case should have it as well, because the whole idea with use cases are to wrap FluxC calls, not hide them. In specific cases, we might want to remove them from the payloads and the use cases together, which I could totally get behind.

> But I guess this is really more about a personal style.

Right :)

Reply:
ok, you are right, I agree.

> because the whole idea with use cases are to wrap FluxC calls

Offtopic: I don't think we need to limit the use of `use-cases` just to the transformation of FluxC events to coroutines. I think we could theoretically put there any logic. We haven't needed it yet, but we might in the future.

Reply:
>This might not have been the best test to bring it up, I probably should have done it in the verticals PR, because even now when I look at the tests for the Verticals VM, I get really overwhelmed and don't know what's going on. I guess I'll have a better understanding of it when I implement the domains VM unit tests and will be interesting choice between following the verticals example and taking a different approach.

I'm looking forward to see your approach:)!

Reply:
> Offtopic: I don't think we need to limit the use of use-cases just to the transformation of FluxC events to coroutines. I think we could theoretically put there any logic. We haven't needed it yet, but we might in the future.

Sure, as long as we don't make it our goal to hide issues in FluxC in the use-cases, I am fine with that. The only thing I want is to improve the FluxC side when necessary instead of introducing a layer for it.

Reply:
Yep, that's a good point and I agree!